Produce correct answers for Group BY NULL (Option 1)

alamb opened the pull request:
# Which issue does this PR close?
This PR closes https://github.com/apache/arrow-datafusion/issues/781 and https://github.com/apache/arrow-datafusion/issues/782;

Built on https://github.com/apache/arrow-datafusion/pull/786 so review that first.

 # Rationale for this change
1. Grouping on columns that contain NULLs today produces incorrect results
2. This is what I think is the minimum change required to produce correct results

This is a version of the "Alternative" approach described in https://github.com/apache/arrow-datafusion/issues/790 which I think is the minimum change to GroupByHash to produce the correct answers when grouping on columns that contain nulls. Thanks to @jhorstmann and @Dandandan  for the ideas leading to this PR

It will likely reduce the speed of grouping as well as require more memory than the current implementation (though it does get correct answers!)

I created this PR  it available for comparison and as a fallback in case I run into trouble or run out of time time trying to implement https://github.com/apache/arrow-datafusion/issues/790, which I expect will take longer to code and review.

# What changes are included in this PR?
1. https://github.com/apache/arrow-datafusion/pull/786
2. Include a "null byte" for each column when creating the key to the hash table
3. Fix some bugs related to NULL handling in `ScalarValue`
4. Tests

On master keys are created like this:
```
                            string len                   0x1234
{                          (as usize le)      "foo"    (as u16 le)
  k1: "foo"         ┌──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┐
  k2: 0x1234u16     │03│00│00│00│00│00│00│00│"f│"o│"o│34│12│
}                   └──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┘
                    0  1  2  3  4  5  6  7  8  9  10 11 12
```

After this PR, the keys are created as follows (note the two extra bytes, one for each grouping column)

Example of a key without any nulls:

```
                       0xFF byte at the start of each column
                          signifies the value is non-null
                                         │

                     ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┴ ─ ─ ─ ─ ─ ─ ─ ┐

                     │        string len                 │  0x1234
{                    ▼       (as usize le)      "foo"    ▼(as u16 le)
  k1: "foo"        ╔ ═┌──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──╦ ═┌──┬──┐
  k2: 0x1234u16     FF║03│00│00│00│00│00│00│00│"f│"o│"o│FF║34│12│
}                  ╚ ═└──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──╩ ═└──┴──┘
                    0  1  2  3  4  5  6  7  8  9  10 11 12 13 14
```

Example of a key with NULL values:
```
                        0xFE byte at the start of k1 column
                    ┌ ─     signifies the value is NULL

                    └ ┐
                             0x1234
{                     ▼    (as u16 le)
  k1: NULL          ╔ ═╔ ═┌──┬──┐
  k2: 0x1234u16      FE║FF║12│34│
}                   ╚ ═╚ ═└──┴──┘
                      0  1  2  3
```


# Are there any user-facing changes?
Correct answers! 

# Benchmark results

The benchmarks show a slight slowdown, which is not unexpected given there is now more work being done

```
group                                                gby_null_alternative                   master
-----                                                --------------------                   ------
aggregate_query_group_by                             1.16      3.6±0.14ms        ? ?/sec    1.00      3.1±0.10ms        ? ?/sec
aggregate_query_group_by_u64 15 12                   1.06      3.7±0.09ms        ? ?/sec    1.00      3.5±0.04ms        ? ?/sec
aggregate_query_group_by_with_filter                 1.06      2.5±0.05ms        ? ?/sec    1.00      2.3±0.04ms        ? ?/sec
aggregate_query_group_by_with_filter_u64 15 12       1.02      2.4±0.10ms        ? ?/sec    1.00      2.4±0.04ms        ? ?/sec
aggregate_query_no_group_by 15 12                    1.00  1152.2±30.38µs        ? ?/sec    1.00  1155.2±28.60µs        ? ?/sec
aggregate_query_no_group_by_count_distinct_narrow    1.14      6.1±0.14ms        ? ?/sec    1.00      5.4±0.05ms        ? ?/sec
aggregate_query_no_group_by_count_distinct_wide      1.18      8.8±0.30ms        ? ?/sec    1.00      7.4±0.10ms        ? ?/sec
aggregate_query_no_group_by_min_max_f64              1.06  1225.6±27.57µs        ? ?/sec    1.00  1160.5±29.22µs        ? ?/sec
```

Dandandan commented on datafusion/tests/sql.rs:
```diff
@@ -3014,6 +3014,109 @@ async fn query_count_distinct() -> Result<()> {
+        "+-----------------+----+",
+        "| COUNT(UInt8(1)) | c1 |",
+        "+-----------------+----+",
+        "| 1               |    |",
```
👍

alamb commented:
Clippy error is unrelated -- see fix in https://github.com/apache/arrow-datafusion/pull/794

Dandandan commented on datafusion/src/physical_plan/hash_aggregate.rs:
```diff
@@ -644,14 +645,63 @@ fn create_key_for_col(col: &ArrayRef, row: usize, vec: &mut Vec<u8>) -> Result<(
     vec.clear();
     for col in group_by_keys {
-        create_key_for_col(col, row, vec)?
+        if !col.is_valid(row) {
```
Not sure if it makes sense to improve performance here, but an optimization might be to check on `null-count==0` outside of this function to avoid the `is_valid` call and just always add an `0xFF`

alamb replied:
Thank you for the suggestion.

If you don't mind I would like to spend time on https://github.com/apache/arrow-datafusion/issues/790 which, if successful, I expect to significantly remove all this code.

I will attempt to add that optimization at a later date.

Dandandan commented on datafusion/src/scalar.rs:
```diff
@@ -1347,6 +1545,29 @@ mod tests {
+    fn scalar_try_from_dict_datatype() {
+        let data_type =
+            DataType::Dictionary(Box::new(DataType::Int8), Box::new(DataType::Utf8));
+        let data_type = &data_type;
```
🥳

alamb replied:
Amusingly, supporting this behavior ended up causing a test to fail when I brought the code into IOx and I think I traced the problem to an issue in parquet file statistics: https://github.com/apache/arrow-rs/issues/641 🤣  this was not a side effect I had anticipated

Dandandan commented on datafusion/tests/sql.rs:
```diff
@@ -3014,6 +3014,109 @@ async fn query_count_distinct() -> Result<()> {
+        "+-----------------+----+-----+",
+        "| 1               |    |     |",
+        "| 2               |    | bar |",
+        "| 3               | 0  |     |",
```
👍

Dandandan commented on datafusion/src/scalar.rs:
```diff
@@ -86,6 +87,120 @@ pub enum ScalarValue {
+        // any newly added enum variant will require editing this list
+        // or else face a compile error
+        match (self, other) {
+            (Boolean(v1), Boolean(v2)) => v1.eq(v2),
```
You could also use `==` instead?

jhorstmann commented:
Looks good. I was trying to come up with an example where two distinct keys would end up as the same encoding but could not find any because any variable length types include the length prefix. :rocket:

alamb commented:
Thanks for the reviews @Dandandan  and @jhorstmann ! I plan to wait another day or two to see if anyone else has feedback on this approach, but what I am thinking of doing is merging this PR (after addressing comments) so at least we get correct answers and then work on the more sophisticated implementation in parallel. 

> Looks good. I was trying to come up with an example where two distinct keys would end up as the same encoding but could not find any because any variable length types include the length prefix. 🚀

Thanks for double checking -- this also worried me a lot so I am glad to hear someone else did a double check too. 

I convinced myself that since each key had entries for the same columns in the same order, there was no way to concoct the same bytes with different column values

jorgecarleitao reviewed (approved):
Went through this carefully, and it looks great! Great work, @alamb !

alamb commented:
Rebased given that https://github.com/apache/arrow-datafusion/pull/786 is merged, so that this PR just shows the delta now

Dandandan commented on datafusion/src/physical_plan/hash_aggregate.rs:
```diff
@@ -508,7 +511,9 @@ fn dictionary_create_key_for_col<K: ArrowDictionaryKeyType>(
-/// `vec` to be used as a key into the hash map
+/// `vec` to be used as a key into the hash map.
+///
+/// NOTE: This functon does not check col.is_valid(). Caller must do so
```
```suggestion
/// NOTE: This function does not check col.is_valid(). Caller must do so
```

alamb commented:
https://github.com/apache/arrow-datafusion/pull/808 contains the PR that should give us back any performance we lost in this one
